File: src/main/java/de/agilecoders/logback/elasticsearch/actor/ErrorHandler.java
package de.agilecoders.logback.elasticsearch.actor;

import akka.actor.DeadLetter;
import akka.actor.UntypedActor;
import ch.qos.logback.classic.Level;
import ch.qos.logback.classic.spi.ILoggingEvent;
import ch.qos.logback.classic.spi.IThrowableProxy;
import ch.qos.logback.classic.spi.LoggerContextVO;
import de.agilecoders.logback.elasticsearch.Configuration;
import de.agilecoders.logback.elasticsearch.ElasticSearchLogbackAppender;
import org.slf4j.Marker;

import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * handles all actor errors
 *
 * @author miha
 */
public class ErrorHandler extends UntypedActor {
    private final ElasticSearchLogbackAppender appender;
    private final Configuration configuration;

    /**
     * Construct.
     *
     * @param appender      the logback to elasticsearch appender
     * @param configuration the base configuration
     */
    ErrorHandler(ElasticSearchLogbackAppender appender, Configuration configuration) {
        this.appender = appender;
        this.configuration = configuration;
    }

    @Override
    public void onReceive(Object message) throws Exception {
        if (message instanceof Failure) {
            handleFailure((Failure) message);
        } else if (message instanceof DeadLetter) {
            handleDeadLetter((DeadLetter) message);
        } else {
            unhandled(message);
        }
    }

    private void handleDeadLetter(DeadLetter message) {
        if (message.message() instanceof ILoggingEvent) {
            retryOrDiscard(message.message());
        } else {
            appender.addError("can't transfer message: " + message);
        }
    }

    private void handleFailure(Failure failure) {
        if (failure.message() != null) {
            retryOrDiscard(failure.message());
        } else {
            appender.addError(failure.errorMessage(), failure.throwable());
        }
    }

    private void retryOrDiscard(Object message) {
        if (message instanceof RetryLoggingEvent) {
            RetryLoggingEvent event = (RetryLoggingEvent) message;

            if (event.counter() <= configuration.retryCount()) {
                appender.doAppend(RetryLoggingEvent.retry(event));
            } else {
                appender.addError("dropped non-discardable event: " + event.toString());
            }
        } else if (message instanceof ILoggingEvent) {
            ILoggingEvent event = (ILoggingEvent) message;

            if (appender.isDiscardable(event)) {
                appender.addWarn("dropped discardable event: " + event.toString());
            } else {
                appender.doAppend(RetryLoggingEvent.retry(event));
            }
        }
    }

    private static final class RetryLoggingEvent implements ILoggingEvent {
        private final AtomicInteger counter;
        private final ILoggingEvent backingEvent;

        private RetryLoggingEvent(ILoggingEvent backingEvent) {
            this.backingEvent = backingEvent;
            this.counter = new AtomicInteger(1);
        }

        public static ILoggingEvent retry(ILoggingEvent event) {
            if (event instanceof RetryLoggingEvent) {
                ((RetryLoggingEvent) event).counter.incrementAndGet();
                return event;
            }

            return new RetryLoggingEvent(event);
        }

        public int counter() {
            return counter.get();
        }

        @Override
        public String getThreadName() {
            return backingEvent.getThreadName();
        }

        @Override
        public Level getLevel() {
            return backingEvent.getLevel();
        }

        @Override
        public String getMessage() {
            return backingEvent.getMessage();
        }

        @Override
        public Object[] getArgumentArray() {
            return backingEvent.getArgumentArray();
        }

        @Override
        public String getFormattedMessage() {
            return backingEvent.getFormattedMessage();
        }

        @Override
        public String getLoggerName() {
            return backingEvent.getLoggerName();
        }

        @Override
        public LoggerContextVO getLoggerContextVO() {
            return backingEvent.getLoggerContextVO();
        }

        @Override
        public IThrowableProxy getThrowableProxy() {
            return backingEvent.getThrowableProxy();
        }

        @Override
        public StackTraceElement[] getCallerData() {
            return backingEvent.getCallerData();
        }

        @Override
        public boolean hasCallerData() {
            return backingEvent.hasCallerData();
        }

        @Override
        public Marker getMarker() {
            return backingEvent.getMarker();
        }

        @Override
        public Map<String, String> getMDCPropertyMap() {
            return backingEvent.getMDCPropertyMap();
        }

        @Override
        public Map<String, String> getMdc() {
            return backingEvent.getMdc();
        }

        @Override
        public long getTimeStamp() {
            return backingEvent.getTimeStamp();
        }

        @Override
        public void prepareForDeferredProcessing() {
            backingEvent.prepareForDeferredProcessing();
        }

        @Override
        public String toString() {
            return backingEvent.toString();
        }
    }

}

File: src/main/scala/de/agilecoders/logback/elasticsearch/actor/ErrorHandler.scala
package de.agilecoders.logback.elasticsearch.actor

import akka.actor._
import ch.qos.logback.classic.spi.ILoggingEvent
import de.agilecoders.logback.elasticsearch.{LogbackContext, IElasticSearchLogbackAppender, CantSendEvent}
import de.agilecoders.logback.elasticsearch.CantSendEvent
import akka.actor.DeadLetter


object ErrorHandler {

    def props(appender: IElasticSearchLogbackAppender): Props = Props(classOf[ErrorHandler], appender)
}

/**
 * TODO miha: document class purpose
 *
 * @author miha
 */
class ErrorHandler(appender: IElasticSearchLogbackAppender) extends Actor with ActorLogging {
    private[this] var errors = 0
    private[this] var deadLetters = 0
    private[this] var cantSend = 0

    override def preStart() = {
        super.preStart()

        log.info("starting ErrorHandler")
    }

    override def postStop() = {
        log.info(s"shutting down ErrorHandler; errors: $errors; cantSend: $cantSend; deadLetters: $deadLetters")

        super.postStop()
    }

    private[this] def handle(failure: CantSendEvent) = failure.message match {
        case e: ILoggingEvent => if(retryOrDiscard(e)) cantSend -= 1
        case _ => appender.addError(failure.message.toString)
    }

    private[this] def handle(deadLetter: DeadLetter) = deadLetter.message match {
        case e: ILoggingEvent => if(retryOrDiscard(e)) deadLetters -= 1
        case _ => appender.addError(deadLetter.message.toString)
    }

    private[this] def retryOrDiscard(message: ILoggingEvent):Boolean = {
        appender.isDiscardable(message) match {
            case true => log.warning(s"dropped discardable event: $message"); false
            case false => {
                val event = RetryLoggingEvent(message)

                event.limitReached match {
                    case true => log.error(s"dropped non-discardable event: $event"); false
                    case false => appender.doAppend(event); true
                }
            }
        }
    }

    def receive = {
        case f: CantSendEvent => {
            log.debug(s"got an cant send event: ${f.hashCode()}")

            cantSend += 1
            handle(f)
        }
        case d: DeadLetter => {
            log.debug(s"got an dead letter: ${d.hashCode()}")

            deadLetters += 1
            handle(d)
        }

        case p: PoisonPill => log.debug("received poison pill")
        case x => errors += 1; log.debug(s"got an error for: $x")
    }
}

object RetryLoggingEvent {
    def retry(e: ILoggingEvent): ILoggingEvent = e match {
        case r: RetryLoggingEvent => r.incrementCounter()
        case _ => RetryLoggingEvent(e)
    }
}

case class RetryLoggingEvent(e: ILoggingEvent) extends ILoggingEvent {
    private[this] var counter: Int = 1
    private[this] val lock = new Object

    def incrementCounter(): RetryLoggingEvent = {
        lock.synchronized {
                              counter = counter + 1
                              this
                          }
    }

    def limitReached: Boolean = counter >= 3

    def prepareForDeferredProcessing() = e.prepareForDeferredProcessing()

    def getThreadName = e.getThreadName

    def getLevel = e.getLevel

    def getMessage = e.getMessage

    def getArgumentArray = e.getArgumentArray

    def getFormattedMessage = e.getFormattedMessage

    def getLoggerName = e.getLoggerName

    def getLoggerContextVO = e.getLoggerContextVO

    def getThrowableProxy = e.getThrowableProxy

    def getCallerData = e.getCallerData

    def hasCallerData = e.hasCallerData

    def getMarker = e.getMarker

    def getMDCPropertyMap = e.getMDCPropertyMap

    @Deprecated
    def getMdc = e.getMdc

    def getTimeStamp = e.getTimeStamp
}